ompi/instance/instance.c

```diff
@@ -59,7 +59,14 @@
 
 ompi_predefined_instance_t ompi_mpi_instance_null = {{{{0}}}};
 
+#if defined(OPAL_RECURSIVE_MUTEX_STATIC_INIT)
 static opal_recursive_mutex_t instance_lock = OPAL_RECURSIVE_MUTEX_STATIC_INIT;
+#else
+static opal_recursive_mutex_t instance_lock;
+__attribute__((__constructor__)) static void instance_lock_init(void) {
```

Comment:
The `constructor` attribute isn't part of the C standard so we should check if it is actually available. What do we do if it isn't?

Reply:
All platforms supporting `PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP` or `PTHREAD_RECURSIVE_MUTEX_INITIALIZER` remain covered as before. The ones where they are not defined but where we have __constructor__ will be now covered. For the rest of them, one could argue that we could provide support for `.init` but that would be more than what I'm willing to invest in this.

There is a single instance of a statically initialized recursive mutex. It is not clear to me why the instance code needs a recursive mutex, but I'm not going to dig deeper.